### PR TITLE
attempt to make apscli configurable from outside

### PR DESCRIPTION
**Pull request by apgsga-jhe**
We need to write (or re-write) unit test for our saveRevision and retrieveRevision mechanism. We naturally have to test what's getting written within Revisions.json file, and we naturally need a different file for testing as the one in production. Without apscli being configurable, this is rather annoying to achieve this. This pull request is something small to introduce a configuration class which (for now) simply reads property(ies) from an external file. The idea of this external file is to be a DEV file (not an OPS one) embedded with apscli. 
What do you think? Do you agree with the approach? 
If you agree, I would merge this one on the master already, even if the test have not all been written yet. Then we would have an environment which would beready to go further.

### CODE REVIEW

**Comment by apgsga-uge** [comment]
why would you want to splic apscli? - I'd think it is an advantage to have it the way it is...

**Comment by chhex** [comment]
@apgsga-uge Agreed it not something, which is clear cut and also not something which necessarily needs to be done (right now)

The ratio: 
a. "apscli" is for using the devops rest api of the Patch Service Server.
Only dependency Spring Rest Template with a Http client
b. "apsrev" is responsible for retrieving, updating the revisions 
Is called from the Pipeline and the onClone Functionality
No further dependencies
c. "apsdb" is responsible for updating the toStatus in the Db
Dependencies jdbc + Database configuration
d. "apsclone" is called from the Clone process and is responsible for "coordinating" what needs to be done on a clone

All would be installed with the same rpm and invocable as scripts (apscli.sh, apsrev.sh, apsdb.sh, apsclone.sh) from the same directory.

The main advantage breaking down the "monolith" apscli , which currently does all of above, in smaller more manageable , independent pieces: 
Let risk and impact with changes and clear responsiblities.

**Comment by chhex** [comment]
@apgsga-jhe  ConfigSlurper similiar to Spring profiles and also distinguish between different configuration targets : http://mrhaki.blogspot.com/2009/08/grassroots-groovy-configuration-with.html and https://touk.pl/blog/2015/09/20/easy-configuration-usage-with-configslurper/ and

**Comment by apgsga-jhe** [comment]
@chhex : I agree that introducing Spring Config could be a bit heavy ... Seems ConfigSlurper would do it ...

**Comment by apgsga-jhe** [comment]
@chhex  :) :) .... ok .... seems like an agreement ;)

**Comment by apgsga-jhe** [comment]
@chhex : I just committed a version which uses ConfigSlurper. The idea would be that we have one single apscli.properties file containing the configuration for the different environment. When "slurping the configuration", we can mention for which environment we're doing it (production by default), and the corresponding props will be loaded. 
I believe this will fulfill our requirements : 
- having one single configuration file, usable from both Eclipse test and in prod.
- possible to configure in which environment we're running.

What do you think?

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 1]
```diff
@@ -2,9 +2,12 @@ package com.apgsga.patch.service.client
 
 
 import org.codehaus.groovy.runtime.StackTraceUtils
+import org.springframework.context.annotation.ClassPathBeanDefinitionScanner
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Ups, done, will be removed with next commit before merge.

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 2]
```diff
@@ -16,14 +19,29 @@ import groovy.json.JsonOutput
...
+	
+	public static PatchCli create(def environment) {
+		env = environment
+		create()
+	}
 
 	public static PatchCli create() {
-		return new PatchCli()
+
+		def url = getClass().getResource('/apscli.properties')
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Done, i'm now using org.springframework.core.io.ClassPathResource ... will be part of next commit before merging.

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 3]
```diff
@@ -16,14 +19,29 @@ import groovy.json.JsonOutput
 import groovy.json.JsonSlurper
 
 class PatchCli {
+	
+	public static PatchCli create(def environment) {
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Done, will be part of next commit before merging.

File: apg-patch-service-cmdclient/src/main/resources/apscli.properties [hunk 4]
```diff
@@ -0,0 +1,8 @@
+environments {
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/resources/apscli.properties]: No prob, I'll integrate these other properties as well.

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 5]
```diff
@@ -16,14 +19,29 @@ import groovy.json.JsonOutput
...
+		def patchCli = new PatchCli()
+		return patchCli
 	}
 
 	private PatchCli() {
 		super();
 	}
+	
 	def validComponents = ["db", "aps", "mockdb", "nil"]
 	def defaultHost = "localhost:9010"
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Done: new property will be called host.default

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 6]
```diff
@@ -32,7 +50,8 @@ class PatchCli {
 	def validToStates
 	def configDir
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Done: new property will be called config.dir

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 7]
```diff
@@ -32,7 +50,8 @@ class PatchCli {
...
 	def defaultConfig
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Seems like this one is already resolved via ConfigSlurper -> valdidateAndLoadConfigFiles() method,

File: apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy [hunk 8]
```diff
@@ -16,14 +19,29 @@ import groovy.json.JsonOutput
...
+	}
 
 	public static PatchCli create() {
-		return new PatchCli()
+
+		def url = getClass().getResource('/apscli.properties')
+		def file = new File(url.toURI())
+		assert file.exists() : "apscli.properties doesn't exist or is not accessible!"
+		ConfigObject conf = new ConfigSlurper(env).parse(url);
+		final Map config = conf.flatten();
```
[the remark this answers is not in the document]
Comment (apgsga-jhe) [reply apg-patch-service-cmdclient/src/main/groovy/com/apgsga/patch/service/client/PatchCli.groovy]: Yep good point, we don't necessarily need to flatten, i'll parse the properties without doing it.

**Pull request merged**